File: src/main/java/discretemath/bool/expression/compound/CompoundBooleanExpression.java
package discretemath.bool.expression.compound;

import java.util.Map;
import java.util.Objects;

import discretemath.bool.expression.atomic.AtomicBooleanExpression;
import discretemath.bool.expression.BooleanExpression;
import discretemath.bool.expression.atomic.BooleanVariable;
import discretemath.bool.expression.exception.VariableValueNotSpecifiedException;
import discretemath.bool.operator.FunctionallyCompleteOperatorsSet;
import discretemath.bool.operator.Operator;
import discretemath.bool.operator.Operators;

public final class CompoundBooleanExpression implements BooleanExpression {

	private final OperatorNode rootNode;

	public CompoundBooleanExpression(OperatorNode rootNode) {
		this.rootNode = Objects.requireNonNull(rootNode, "Root not may not be null");
	}

	public OperatorNode getRootNode() {
		return rootNode;
	}

	@Override
	public boolean evaluate(Map<BooleanVariable, Boolean> arguments) throws VariableValueNotSpecifiedException {
		return evaluateNode(rootNode, arguments);
	}

	@Override
	public BooleanExpression expressViaOperators(FunctionallyCompleteOperatorsSet operatorsSet) {
		return expressExpressionViaOperators(this.rootNode, operatorsSet);
	}

	@Override
	public BooleanExpression normalize() {
		OperatorNode newRootNode = (OperatorNode) deleteDoubleNegations(rootNode, false);
		return new CompoundBooleanExpression(newRootNode);
	}

	@Override
	public String toString() {
		return rootNode.toString();
	}

	private static boolean evaluateNode(ExpressionTreeNode node, Map<BooleanVariable, Boolean> values) {
		if (node instanceof ValueNode valueNode) {
			AtomicBooleanExpression expression = valueNode.expression();
			return expression.evaluate(values);
		} else {
			OperatorNode operatorNode = (OperatorNode) node;
			Operator operator = operatorNode.operator();
			ExpressionTreeNode[] operands = operatorNode.operands();
			boolean[] operandsValues = new boolean[operands.length];
			for (int i = 0; i < operands.length; i++) {
				ExpressionTreeNode operand = operands[i];

				operandsValues[i] = evaluateNode(operand, values);
			}

			return operator.evaluate(operandsValues);
		}
	}

	private static BooleanExpression expressExpressionViaOperators(ExpressionTreeNode node,
																   FunctionallyCompleteOperatorsSet operatorsSet) {
		if (node instanceof ValueNode valueNode) {
			return valueNode.expression();
		}

		OperatorNode operatorNode = (OperatorNode) node;
		Operator operator = operatorNode.operator();
		ExpressionTreeNode[] operands = operatorNode.operands();

		BooleanExpression[] operandExpressions = ExpressionTreeNode.toBooleanExpression(operands);

		return operator.applyUsing(operatorsSet, operandExpressions);
	}

	private static ExpressionTreeNode deleteDoubleNegations(ExpressionTreeNode node, boolean negationAlreadyFound) {
		if (node instanceof ValueNode) {
			if (negationAlreadyFound) {
				return OperatorNode.forNegation(node);
			} else {
				return node;
			}
		}

		OperatorNode operatorNode = (OperatorNode) node;

		if (operatorNode.operator() == Operators.NOT) {
			ExpressionTreeNode operand = operatorNode.operands()[0];

			if (negationAlreadyFound) {
				return deleteDoubleNegations(operand, false);
			} else {
				return deleteDoubleNegations(operand, true);
			}
		} else {
			ExpressionTreeNode[] operands = operatorNode.operands();
			ExpressionTreeNode[] newOperands = new ExpressionTreeNode[operands.length];
			for (int i = 0; i < operands.length; i++) {
				ExpressionTreeNode operand = operands[i];
				newOperands[i] = deleteDoubleNegations(operand, false);
			}

			OperatorNode newOperatorNode = new OperatorNode(operatorNode.operator(), newOperands);
			if (negationAlreadyFound) {
				return OperatorNode.forNegation(newOperatorNode);
			} else {
				return newOperatorNode;
			}
		}
	}
}

File: src/main/scala/discretemath/bool/expression/compound/CompoundBooleanExpression.scala
package discretemath.bool.expression.compound

import discretemath.bool.expression.BooleanExpression
import discretemath.bool.expression.atomic.BooleanVariable
import discretemath.bool.expression.compound.CompoundBooleanExpression.collectNodeVariables
import discretemath.bool.expression.exception.VariableValueNotSpecifiedException
import discretemath.bool.operator.FunctionallyCompleteOperatorsSet
import discretemath.bool.operator.Operators.NOT
import discretemath.combination.BitCombinations

import scala.collection.mutable

final class CompoundBooleanExpression(val rootNode: OperatorNode) extends BooleanExpression {
  assert(rootNode != null, "Root not may not be null")

  override def getVariables: Set[BooleanVariable] = {
    val allVariables = new mutable.HashSet[BooleanVariable]()
    collectNodeVariables(rootNode, allVariables)

    allVariables.toSet // TODO redundant copy for immutable set
  }

  @throws[VariableValueNotSpecifiedException]
  override def evaluate(arguments: Map[BooleanVariable, Boolean]): Boolean = CompoundBooleanExpression.evaluateNode(rootNode, arguments)

  override def expressViaOperators(operatorsSet: FunctionallyCompleteOperatorsSet): BooleanExpression = CompoundBooleanExpression.expressNodeViaOperators(this.rootNode, operatorsSet)

  override def minimize: BooleanExpression = {
    val newRootNode = CompoundBooleanExpression.deleteDoubleNegations(rootNode, negationAlreadyFound = false).asInstanceOf[OperatorNode]
    new CompoundBooleanExpression(newRootNode)
  }

  override def isSatisfiable: Boolean = {
    val variablesList = getVariables.toIndexedSeq
    val allPossibleValues = BitCombinations.nBitCombinations(variablesList.size)


    for (i <- 0 until allPossibleValues.rowsCount) {
      val values = allPossibleValues.getRow(i)

      val variableMap = variablesList.zipWithIndex.map(tuple => (tuple._1, values(tuple._2))).toMap

      val result = evaluate(variableMap)

      if (result) {
        return true
      }
    }

    return false
  }

  override def toString: String = rootNode.toString
}

object CompoundBooleanExpression {

  def apply(rootNode: OperatorNode): CompoundBooleanExpression = new CompoundBooleanExpression(rootNode)

  private def evaluateNode(node: ExpressionTreeNode, values: Map[BooleanVariable, Boolean]): Boolean = {
    node match {
      case ValueNode(expression) => expression.evaluate(values)
      case OperatorNode(operator, operands) => operator.evaluate(operands.map(evaluateNode(_, values)))
    }
  }

  private def collectNodeVariables(node: ExpressionTreeNode, variables: mutable.Set[BooleanVariable]): Unit = {
    node match {
      case ValueNode(expression) => variables.addAll(expression.getVariables)
      case OperatorNode(_, operands) => operands.foreach(operand => collectNodeVariables(operand, variables))
    }
  }

  private def expressNodeViaOperators(node: ExpressionTreeNode, operatorsSet: FunctionallyCompleteOperatorsSet): BooleanExpression = {
    node match {
      case ValueNode(expression) => expression.expressViaOperators(operatorsSet)
      case OperatorNode(operator, operands) => operator.injectUsing(operatorsSet, ExpressionTreeNode.toBooleanExpression(operands))
    }
  }

  private def deleteDoubleNegations(node: ExpressionTreeNode, negationAlreadyFound: Boolean): ExpressionTreeNode = {
    node match {
      case ValueNode(_) => if (negationAlreadyFound) OperatorNode.not(node) else node
      case OperatorNode(operator, operands) if operator == NOT => {
        val operand = operands.head
        if (negationAlreadyFound) deleteDoubleNegations(operand, negationAlreadyFound = false)
        else deleteDoubleNegations(operand, negationAlreadyFound = true)
      }
      case OperatorNode(operator, operands) => {
        val newOperands = operands.map(operand => deleteDoubleNegations(operand, negationAlreadyFound = false))

        val newOperatorNode = OperatorNode(operator, newOperands)
        return if (negationAlreadyFound) OperatorNode.not(newOperatorNode) else newOperatorNode
      }
    }
  }
}